### PR TITLE
Add HCR mechanisms.

### PR DESCRIPTION
**Pull request by yaohaizh**
@andxu @testforstephen @akaroml @ansyral

### CODE REVIEW

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 1]
```diff
@@ -240,58 +265,386 @@ public HotCodeReplaceEvent(String message) {
...
+                context.getProtocolServer().sendEvent(new HotCodeReplaceEvent("JVM contains obsolete methods"));
+            }
+
+            if (currentDebugSession.getVM().canPopFrames() && framesPopped) {
+                attemptStepIn(poppedThreads);
+            } else {
+                attemptDropToFrame(resourcesToReplace, qualifiedNamesToReplace);
+            }
+        } catch (DebugException e) {
+            logger.log(Level.SEVERE, "Failed to compolete hot code replace: " + e.getMessage(), e);
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: complete

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 2]
```diff
@@ -343,4 +696,37 @@ public static IJavaElement findElement(String qualifiedTypeName, IJavaProject pr
...
+            for (StackFrame frame : frames) {
+                if (StackFrameUtility.isObsolete(frame)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private List<StackFrame> getStackFrames(ThreadReference thread, boolean refresh) {
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: use compute/computeIfAbsent

Comment (andxu) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: threadFrameMap.compute(thread, (key, old) -> {
            try {
                return old == null || refresh ? thread.frames() : old;
            } catch (IncompatibleThreadStateException e) {
                logger.log(Level.SEVERE, "Failed to get stack frames: " + e.getMessage(), e);
                return old;
            }
        });

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 3]
```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
...
+            }
+        }
+        return false;
+    }
+
+    private String getDeclaringTypeName(StackFrame frame) throws DebugException {
+        return getGenericName(StackFrameUtility.getDeclaringType(frame));
+    }
+
+    private String getGenericName(ReferenceType type) throws DebugException {
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: I remember @andxu implemented the similar format function in variable handler.  is there any possible to reuse?

Comment (yaohaizh) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: @andxu's change is not suitable here.  And he discarded his change.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/IHotCodeReplaceProvider.java [hunk 4]
```diff
@@ -12,8 +12,8 @@
 package com.microsoft.java.debug.core.adapter;
 
 import java.util.List;
-import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
 
 public interface IHotCodeReplaceProvider extends IProvider {
-    CompletableFuture<List<String>> redefineClasses();
+    void redefineClasses(Consumer<List<String>> consumer);
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/IHotCodeReplaceProvider.java]: In the new design, this is more like an event. Please define the interface in event fashion.
```java
IDisposable onClassRedefined(Consumer callback);
```

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 5]
```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
...
             }
         }
     }
 
     @Override
-    public CompletableFuture<List<String>> redefineClasses() {
-        return CompletableFuture.supplyAsync(() -> {
-            synchronized (needHotCodeReplace) {
+    public void redefineClasses(Consumer<List<String>> consumer) {
+        this.consumer = consumer;
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: The event may take multiple subscribers and allow unsubscribing. RXJava interface is a good reference.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 6]
```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
...
+        }
+    }
+
+    protected void dropToStackFrame(StackFrame frame) throws DebugException {
+        // Pop the drop frame and all frames above it
+        popStackFrame(frame);
+        stepIntoThread(frame.thread());
+    }
+
+    private void redefineTypesJDK(List<IResource> resources, List<String> qualifiedNames) throws DebugException {
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: Why is the function name ends with JDK?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 7]
```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
...
+    }
+
+    /**
+     * Returns a list of frames which should be popped in the given threads.
+     */
+    protected List<StackFrame> getAffectedFrames(List<ThreadReference> threads, List<String> replacedClassNames)
+            throws DebugException {
+        List<StackFrame> popFrames = new ArrayList<>();
+        for (ThreadReference thread : threads) {
+            if (thread.isSuspended()) {
```
[the remark this answers is not in the document]
Comment (yaohaizh) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: If the thread is not suspended, we will avoid do any operation on the thread.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java (hunk 6 above)
Comment (yaohaizh) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: There is other redefinTypes other than the JDK version, such as IBM's J9. Just use this naming convention from JDT's implementation.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 8]
```diff
@@ -64,7 +83,13 @@
 
     private static final String CLASS_FILE_EXTENSION = "class"; //$NON-NLS-1$
 
-    private AtomicBoolean needHotCodeReplace = new AtomicBoolean(false);
+    private IDebugSession currentDebugSession;
+
+    private IDebugAdapterContext context;
+
+    private Map<ThreadReference, List<StackFrame>> threadFrameMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (yaohaizh) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: Right. Andy and I had talked about this. And we will guard this from the context side when both side code is ready for merge.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java [hunk 9]
```diff
@@ -240,56 +265,384 @@ public HotCodeReplaceEvent(String message) {
...
+        }
+
+        filterNotLoadedTypes(resourcesToReplace, qualifiedNamesToReplace);
+        if (qualifiedNamesToReplace.isEmpty()) {
+            return;
+            // If none of the changed types are loaded, do nothing.
+        }
+
+        // Not supported scenario:
+        if (!currentDebugSession.getVM().canRedefineClasses()) {
```
[the remark this answers is not in the document]
Comment (yaohaizh) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: This is a characteristic of JVM.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java (hunk 2 above)
Comment (yaohaizh) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java]: Fixed.

**Pull request merged**